src/Core/VarReplay/varinfo.jl

```diff
@@ -0,0 +1,325 @@
+###########
+# VarName #
+###########
+struct VarName{sym}
+    csym      ::    Symbol        # symbol generated in compilation time
+    indexing  ::    String        # indexing
+    counter   ::    Int           # counter of same {csym, uid}
+end
+VarName(csym, sym, indexing, counter) = VarName{sym}(csym, indexing, counter)
+
+function Base.getproperty(vn::VarName{sym}, f::Symbol) where {sym}
+    return f === :sym ? sym : getfield(vn, f)
+end
+
+# NOTE: VarName should only be constructed by VarInfo internally due to the nature of the counter field.
+
+uid(vn::VarName) = (vn.csym, vn.sym, vn.indexing, vn.counter)
+Base.hash(vn::VarName) = hash(uid(vn))
+
+isequal(x::VarName, y::VarName) = hash(uid(x)) == hash(uid(y))
+==(x::VarName, y::VarName)      = isequal(x, y)
+
+Base.string(vn::VarName) = "{$(vn.csym),$(vn.sym)$(vn.indexing)}:$(vn.counter)"
+Base.string(vns::Vector{<:VarName}) = replace(string(map(vn -> string(vn), vns)), "String" => "")
+
+sym(vn::VarName) = Symbol("$(vn.sym)$(vn.indexing)")  # simplified symbol
+
+cuid(vn::VarName) = (vn.csym, vn.sym, vn.indexing)    # the uid which is only available at compile time
+
+copybyindex(vn::VarName, indexing::String) = VarName(vn.csym, vn.sym, indexing, vn.counter)
+
+###########
+# VarInfo #
+###########
+
+abstract type AbstractVarInfo end
+const VarInfo = AbstractVarInfo
+
+mutable struct UntypedVarInfo <: AbstractVarInfo
+    idcs        ::    Dict{VarName,Int}
+    vns         ::    Vector{VarName}
+    ranges      ::    Vector{UnitRange{Int}}
+    vals        ::    Vector{Real}
+    dists       ::    Vector{Distributions.Distribution}
+    gids        ::    Vector{Int}
+    logp        ::    Real
+    num_produce ::    Int           # num of produce calls from trace, each produce corresponds to an observe.
+    orders      ::    Vector{Int}   # observe statements number associated with random variables
+    flags       ::    Dict{String, BitVector}
```

Comment:
If this `flags` filed leads to instability, we can also ``unpack'' it into several fields. The main purpose of it is to introduce an unified interface for flag-like fields.

Reply:
No this is fine, since the way we use it is type stable as far as I can tell.